openquake/hazardlib/gsim/campbell_1997.py

```diff
@@ -0,0 +1,145 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with OpenQuake. If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module exports :class:`Campbell1997`
+"""
+import numpy as np
+
+from openquake.hazardlib.gsim.base import GMPE
+from openquake.hazardlib import const
+from openquake.hazardlib.imt import PGA
+
+
+class Campbell1997(GMPE):
+    """
+    Implements GMPE (PGA) by Campbell, Kenneth W. "Empirical near-source
+    attenuation relationships for horizontal and vertical components of peak
+    ground acceleration, peak ground velocity, and pseudo-absolute acceleration
+    response spectra." Seismological research letters 68.1 (1997): 154-179.
+    """
+    #: Supported TRT active...we specify active_shallow_crust
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Supported intensity measure types are PGA, PGV, PSA, but we only define
+    #: PGA because this is the only IMT used by an implemented model (09/18)
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([
+        PGA
+    ])
+
+    #: Supported intensity measure component is the horizontal component
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.AVERAGE_HORIZONTAL
+
+    #: Supported standard deviation type is only total, see equation 4, pg 164
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([
+        const.StdDev.TOTAL,
+    ])
+
+    #: Requires vs30
+    REQUIRES_SITES_PARAMETERS = set(('vs30',))
+
+    #: Required rupture parameters are magnitude and top of rupture depth
+    REQUIRES_RUPTURE_PARAMETERS = set(('mag', 'rake'))
+
+    #: Required distance measure is closest distance to rupture
+    REQUIRES_DISTANCES = set(('rrup', ))
+
+    #: Verification of the mean value was done by digitizing Figs. 9 and 10
+    #: using Engauge Digitizer. The tests check varied magnitude, distance,
+    #: vs30, and faulting type. Maximum error was ~1.3%. OpenQuake trellis
+    #: plots match these figures. Also tested against a matlab implementation
+    #: (web.stanford.edu/~bakerjw/GMPEs/C_1997_horiz.m), which also has no
+    #: verification tables.
+
+    non_verified = True
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """
+        See :meth:`superclass method
+        <.base.GroundShakingIntensityModel.get_mean_and_stddevs>`
+        for spec of input and result values.
+        """
+        R = (dists.rrup)
+        M = rup.mag
+        # get constants
+        Ssr = self.get_Ssr_term(sites.vs30)
+        Shr = self.get_Shr_term(sites.vs30)
+        rake = rup.rake
+        F = self.get_fault_term(rake)
+
+        # compute mean
+        mean = -3.512 + (0.904 * M) - (1.328 * np.log(np.sqrt(R**2
+               + (0.149 * np.exp(0.647 * M))**2))) \
+               + (1.125 - 0.112 * np.log(R) - 0.0957 * M) * F \
+               + (0.440 - 0.171 * np.log(R)) * Ssr \
+               + (0.405 - 0.222 * np.log(R)) * Shr
+        stddevs = self.get_stddevs(mean)
```

Comment:
I would make two notes here for general implementation of GMPEs, that are not necessarily fundamental to this one (as this is legacy rather than current best practice) but can help significantly in keeping the GMPE library consistent (which is better for the GMPE toolkit or secondary uses of the library):

1. It is more useful to split the GMPE into separate functions for the different elements of the model (i.e. magnitude scaling, geometric spreading, anelastic attenuation, depth/hanging wall/style-of-faulting adjustments and site amplification) rather than execute the entire function in one continuous line. This makes it easier to spawn modified/adjusted GMPEs in subclasses that adapt certain parts of the model (e.g. changing the anelastic term for application to a different region). This is becoming more important to keep doing now than ever before, as it is more common in best practice PSHA to see "backbone" GMPEs adjusted for local conditions and applications than it is to see new GMPEs for different regions/applications.

2. Even though this GMPE is only for PGA I would maintain it is still better to use a coefficients table and place the coefficients there rather than coding them into the call to the main function itself. The coefficients table is another way of helping users make modifications, check things like spectral ranges (though that's easily side-stepped here), and make on-the-fly adjustments to the models (yes, that's possible!) without having to re-write too many functions.